ytypes/node.go

```diff
@@ -186,40 +188,47 @@ func retrieveNodeList(schema *yang.Entry, root interface{}, path, traversedPath
 
 		// Handle lists with a single key.
 		if !util.IsValueStruct(k) {
-			// Handle the special case that we have zero keys specified only when we are handling lists
-			// with partial keys specified.
-			if len(path.GetElem()[0].GetKey()) == 0 && args.partialKeyMatch {
-				keys, err := ygot.PathKeyFromStruct(listElemV)
-				if err != nil {
-					return nil, status.Errorf(codes.Unknown, "could not get path keys at %v: %v", traversedPath, err)
+			match := false
+			switch {
+			case len(path.GetElem()[0].GetKey()) == 0 && args.partialKeyMatch:
+				// Handle the special case that we have zero keys specified only when we are handling lists
+				// with partial keys specified.
+				match = true
+			default:
+				// Otherwise, check for equality of the key.
+				pathKey, ok := path.GetElem()[0].GetKey()[schema.Key]
+				if !ok {
+					return nil, status.Errorf(codes.NotFound, "schema key %s is not found in gNMI path %v, root %T", schema.Key, path, root)
 				}
-				nodes, err := retrieveNode(schema, listElemV.Interface(), util.PopGNMIPath(path), appendElem(traversedPath, &gpb.PathElem{Name: path.GetElem()[0].Name, Key: keys}), args)
+
+				kv, err := getKeyValue(listElemV.Elem(), schema.Key)
 				if err != nil {
-					return nil, err
+					return nil, status.Errorf(codes.Unknown, "failed to get key value for %v, path %v: %v", listElemV.Interface(), path, err)
 				}
 
-				matches = append(matches, nodes...)
-
-				continue
-			}
+				keyAsString, err := ygot.KeyValueAsString(kv)
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "failed to convert %v to a string, path %v: %v", kv, path, err)
+				}
 
-			// Otherwise, check for equality of the key.
-			pathKey, ok := path.GetElem()[0].GetKey()[schema.Key]
-			if !ok {
-				return nil, status.Errorf(codes.NotFound, "schema key %s is not found in gNMI path %v, root %T", schema.Key, path, root)
+				if (args.handleWildcards && pathKey == "*") || keyAsString == pathKey {
+					match = true
+				}
 			}
+			if match {
+				keys, err := ygot.PathKeyFromStruct(listElemV)
+				// if err != nil {
+				// 	return nil, status.Errorf(codes.Unknown, "could not get path keys at %v: %v", traversedPath, err)
+				// }
```

Comment:
There is a Problem with this error check.
If it turned on, we get errors in other tests.
The problem ist, the code hasn't be there before refactoring.

Reply:
Can we leave it here please -- we need to check against `master` to see if something has changed. I'll review this.